src/irmin-pack/unix/gc_worker.ml

```diff
@@ -267,57 +278,82 @@ module Make (Args : Gc_args.S) = struct
         (Commit_value.parents commit)
     in
 
-    (* Step 6. Create the new suffix and prepare 2 functions for read and write
-       operations. *)
-    stats := Gc_stats.Worker.finish_current_step !stats "suffix: start";
-    [%log.debug "GC: creating new suffix"];
-    let suffix = create_new_suffix ~root ~generation in
-    Errors.finalise_exn (fun _outcome ->
-        Ao.fsync suffix
-        >>= (fun _ -> Ao.close suffix)
-        |> Errs.log_if_error "GC: Close suffix")
-    @@ fun () ->
-    let append_exn = Ao.append_exn suffix in
-
-    (* Step 7. Transfer to the next suffix. *)
-    [%log.debug "GC: transfering to the new suffix"];
-    stats := Gc_stats.Worker.finish_current_step !stats "suffix: transfer";
-    let num_iterations = 7 in
-    (* [transfer_loop] is needed because after garbage collection there may be new objects
-       at the end of the suffix file that need to be copied over *)
-    let rec transfer_loop i ~off =
-      if i = 0 then off
-      else
-        let () = Fm.reload fm |> Errs.raise_if_error in
-        let pl : Payload.t = Fm.Control.payload (Fm.control fm) in
-        let end_offset =
-          Dispatcher.offset_of_suffix_poff dispatcher pl.suffix_end_poff
-        in
-        let len = Int63.Syntax.(end_offset - off) in
-        [%log.debug
-          "GC: transfer_loop iteration %d, offset %a, length %a"
-            (num_iterations - i + 1)
-            Int63.pp off Int63.pp len];
-        stats := Gc_stats.Worker.add_suffix_transfer !stats len;
-        let () = Dispatcher.read_bytes_exn dispatcher ~f:append_exn ~off ~len in
-        (* Check how many bytes are left, [4096*5] is selected because it is roughly the
-           number of bytes that requires a read from the block device on ext4 *)
-        if Int63.to_int len < 4096 * 5 then end_offset
-        else
-          let off = Int63.Syntax.(off + len) in
-          transfer_loop ~off (i - 1)
-    in
-    let new_end_suffix_offset =
-      transfer_loop ~off:new_prefix_end_offset num_iterations
+    (* Step 6. Calculate new suffix values. *)
+    let new_suffix, chunk_idx_gced =
+      stats :=
+        Gc_stats.Worker.finish_current_step !stats
+          "suffix: calculate new values";
+      let suffix = Fm.suffix fm in
+      let soff =
+        Dispatcher.suffix_poff_of_offset dispatcher new_suffix_start_offset
+      in
+      (* Step 6.1. Calculate chunks that we have GCed. *)
+      let open struct
+        type chunk = { idx : int; end_suffix_off : int63 }
+      end in
+      let chunks_gced =
+        match Fm.Suffix.chunk_num suffix with
+        | 1 -> [] (* We never GC a single chunk. *)
+        | _ ->
+            Fm.Suffix.fold_chunks
```

Comment:
It seems to me we only need the idx and end_suffix_offset of the last chunk. If we assume that the chunks ids are in order, and we know the initial start_chunk_idx, I think we can compute everything else from that?

Reply:
Hmm, I think the parameters are the smallest set to achieve the above (and also felt reasonable for a general fold of chunks), but it's possible I'm missing a simplification.

Here are some reasons I went this path with the code:
1. I didn't want to have logic about chunks (eg `is_appendable`) outside of `Chunked_suffix`
2. I wanted the logic for deciding if a chunk is removed or not to be self-contained in the function
3. I didn't want to expose the internal `chunk` type or a new type just for this (although I was tempted given the parameter number)
4. I didn't want to encode any assumptions about ordering of chunks or ids (sort of related to 1.).
5. For code with conditionals like this, I like to encode the boolean in as readable language as possible, to aid in understanding for future readers of the code (like myself :sweat_smile:).

WDYT?

Reply:
I haven't spend as much time as you on this code, so its not very clear to me what type of info on the chunks are fine to be manipulated in this file and which are not. For instance, Fm.Suffix chould return the chunk_idx and end_suffix_off of the last chunk to be removed, based on the `soff`. We can do the check that it is the not appendable or empty chunk only on this chunk (as we anyway won't have multiple appendable or empty chunks).

Why I find this slightly unintuitive is that in practice, we will remove one chunk file per gc. Not saying we should only code for that case, obviously. 

(I find it interesting to discuss design choices, thus my lengthy response, but feel free to not go with my suggestions:) )

Reply:
> For instance, Fm.Suffix chould return the chunk_idx and end_suffix_off of the last chunk to be removed, based on the soff. 

This is an interesting idea. If I am following what you're saying, another way to write this code is to add a function to `Chunked_suffix` that returns "removable" chunks based on a `soff`. I did briefly consider that option but ultimately did not choose it because I wanted the "removable" logic (in our situation, the GC logic) to live in `gc_worker.ml` so that a reader can scan one place for all of the core GC logic. If I didn't have this particular use case for the code, I might have chosen to add that to `Chunked_suffix`'s API though.

Add this to what I said previously, I feel good about the current construction/division of the code.

> Why I find this slightly unintuitive is that in practice, we will remove one chunk file per gc.

Do you think this code would read as unintuitive if you came to it fresh without thinking about that (since as you mention, it isn't necessarily the case)?

Reply:
There is not a clear way to make the code intuitive on a first read anyway. Your choices seem good to me, thanks for the explanations.